### PR TITLE
Ported zero copy patch from DAOS v1.0 to v1.2

### PR DESCRIPTION
**Pull request by igalanis**
Signed-off-by: Ioannis Galanis <ioannis.galanis@intel.com>

### CODE REVIEW

File: src/bio/bio_buffer.c [hunk 1]
```diff
@@ -851,10 +854,16 @@ copy_one(struct bio_desc *biod, struct bio_iov *biov,
...
-			bio_memcpy(biod, media, addr, iov->iov_buf +
-					arg->ca_iov_off, nob);
+			if (arg->ca_zc_fetch) {
+				/* just use the DMA buffer */
+				iov->iov_buf = addr;
+			} else {
+				bio_memcpy(biod, media, addr,
+					   iov->iov_buf +
+					   arg->ca_iov_off, nob);
+			}
```
[the remark this answers is not in the document]
Comment (gnailzenh) [reply src/bio/bio_buffer.c:866]: Please fix the style issue

File: src/bio/bio_buffer.c [hunk 2]
```diff
@@ -912,12 +921,14 @@ bio_iod_prep(struct bio_desc *biod, unsigned int type)
 
 	biod->bd_chk_type = type;
 retry:
+
```
Comment (gnailzenh) [inline src/bio/bio_buffer.c:924]: please revert unnecessary change and keep the patch clean.

File: src/bio/bio_buffer.c [hunk 3]
```diff
@@ -912,12 +921,14 @@ bio_iod_prep(struct bio_desc *biod, unsigned int type)
...
 	rc = iterate_biov(biod, dma_map_one, NULL);
 	if (rc) {
 		/*
 		 * To avoid deadlock, held buffers need be released
 		 * before waiting for other active IODs.
 		 */
+		D_DEBUG(DB_IO, "here");
```
Comment (gnailzenh) [inline src/bio/bio_buffer.c:931]: please revert unnecessary change and keep the patch clean.

File: src/bio/bio_buffer.c [hunk 4]
```diff
@@ -943,8 +954,10 @@ bio_iod_prep(struct bio_desc *biod, unsigned int type)
 
 		goto retry;
 	}
-	biod->bd_buffer_prep = 1;
 
+
+	biod->bd_buffer_prep = 1;
+	D_DEBUG(DB_IO, "biod->bd_buffer_prep %d", biod->bd_buffer_prep);
```
Comment (gnailzenh) [inline src/bio/bio_buffer.c:960]: This debug message is not quite helpful

File: src/bio/bio_buffer.c [hunk 5]
```diff
@@ -975,12 +989,17 @@ int
 bio_iod_post(struct bio_desc *biod)
 {
 	struct bio_dma_buffer *bdb;
+	D_DEBUG(DB_IO,"biod->bd_buffer_prep %d", (int)biod->bd_buffer_prep);
 
 	if (!biod->bd_buffer_prep)
-		return -DER_INVAL;
+		{
```
Comment (gnailzenh) [inline src/bio/bio_buffer.c:995]: style: 
```
if (!boid->bd_buffer_prep) {
}
```

**Review by gnailzenh — COMMENTED** [review]
Hi Ioannis, I think we should try to keep the branch clean and neat, so you might have to remove all commented code, remove those temporary debug information, and revert unnecessary changes.

File: src/bio/bio_buffer.c (hunk 4 above)
Comment (igalanis) [reply src/bio/bio_buffer.c:960]: removed the message.

File: src/bio/bio_buffer.c [hunk 6]
```diff
@@ -975,12 +989,17 @@ int
 bio_iod_post(struct bio_desc *biod)
 {
 	struct bio_dma_buffer *bdb;
+	D_DEBUG(DB_IO,"biod->bd_buffer_prep %d", (int)biod->bd_buffer_prep);
```
[the remark this answers is not in the document]
Comment (igalanis) [reply src/bio/bio_buffer.c:992]: removed the debug message

File: src/bio/bio_buffer.c [hunk 7]
```diff
@@ -989,7 +1008,7 @@ bio_iod_post(struct bio_desc *biod)
 		dma_rw(biod, false);
 	else
 		biod->bd_result = 0;
-
+	D_DEBUG(DB_IO,"biod->bd_result %d", (int)biod->bd_result);
```
[the remark this answers is not in the document]
Comment (igalanis) [reply src/bio/bio_buffer.c:1011]: removed the debug message.

File: src/bio/bio_buffer.c [hunk 8]
```diff
@@ -1010,8 +1030,14 @@ bio_iod_copy(struct bio_desc *biod, d_sg_list_t *sgls, unsigned int nr_sgl)
...
-
-	return iterate_biov(biod, copy_one, &arg);
+	arg.ca_zc_fetch = zc_fetch;
+	
+	int rc = iterate_biov(biod, copy_one, &arg);
+	d_iov_t *iov;
+	iov = arg.ca_sgls->sg_iovs;
+	D_DEBUG(DB_IO, "buf %s", (char *)iov->iov_buf);
+	return rc;
+
```
[the remark this answers is not in the document]
Comment (igalanis) [reply src/bio/bio_buffer.c:1040]: removed the message to keep the code clean.

File: src/object/obj_rpc.c [hunk 9]
```diff
@@ -1143,12 +1143,15 @@ int
...
 	void *reply = crt_reply_get(rpc);
-
+	
 	switch (opc_get(rpc->cr_opc)) {
 	case DAOS_OBJ_RPC_UPDATE:
 	case DAOS_OBJ_RPC_TGT_UPDATE:
 	case DAOS_OBJ_RPC_FETCH:
-		return ((struct obj_rw_out *)reply)->orw_ret;
+		{
+			D_DEBUG(DB_IO,"crt_reply: %d, DAOS_OBJ_RPC_UPDATE %d", (int) opc_get(rpc->cr_opc), (int)DAOS_OBJ_RPC_UPDATE);
```
[the remark this answers is not in the document]
Comment (igalanis) [reply src/object/obj_rpc.c:1152]: deleted debug message.

File: src/object/srv_obj.c [hunk 10]
```diff
@@ -82,16 +82,37 @@ obj_gen_dtx_mbs(struct daos_shard_tgt *tgts, bool is_ec, uint32_t *tgt_cnt,
...
-	int			 rc;
 
-	if (daos_handle_is_valid(ioh)) {
+	// if (daos_handle_is_valid(ioh)) {
+	if (!daos_handle_is_inval(ioh)) {
 		bool update = obj_rpc_is_update(rpc);
 
+		int rc;
+		uint32_t map_version;
+		if (ioc->ioc_zc_fetch) {
```
[the remark this answers is not in the document]
Comment (igalanis) [reply src/object/srv_obj.c:97]: fixed
